### PR TITLE
[disruption budget controller] implement fast ResourceEventHandler

### PR DESCRIPTION
**Pull request by veshij**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Existing implementation of ResourceEventHandlerFuncs handling pod updates in pod disruption budget controller could be relatively slow (getPdbForPod() is expensive) under certain conditions (see linked issue).
It leads to unbounded pendingNotifications buffer growth (see memory profiles in linked issue): https://sourcegraph.com/github.com/kubernetes/kubernetes@94a15929cf13354fdf3747cb266d511154f8c97b/-/blob/staging/src/k8s.io/client-go/tools/cache/shared_informer.go?L872-877).
Since it happens before the disruption budget controller takes action - we don't see elevated queue/processing times for PDB workqueues.

Updating the code to make ResourceHandlerFunctions fast and mostly non-blocking.
* extract labels and discard original pod object immediately after receiving in add/update/delete ResourceHandlerFunctions. 
* introduce new queue which processes individual pod labels. As a side-effect - we deduplicate events for the same PodDisruptionBudget (if many pods for the same PDB is updated in the queue - we recalculate PDB state only once).
* it also fixes a potential issue with old PDB not being updated if pod changes labels: https://sourcegraph.com/github.com/kubernetes/kubernetes@94a1592/-/blob/pkg/controller/disruption/disruption.go?L494-507 

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes/kubernetes/issues/117123

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/listers/policy/v1/poddisruptionbudget_expansion.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ import (
 // PodDisruptionBudgetLister.
 type PodDisruptionBudgetListerExpansion interface {
 	GetPodPodDisruptionBudgets(pod *v1.Pod) ([]*policy.PodDisruptionBudget, error)
```
Comment (veshij) [inline staging/src/k8s.io/client-go/listers/policy/v1/poddisruptionbudget_expansion.go:31]: Not used, but I'm not sure if could be safely deleted.

**Comment by veshij** [comment]
@mortent @soltysh @liggitt sorry for pinging you directly folks.
Is there anything I'm missing to get this PR reviewed?

File: pkg/controller/disruption/disruption.go [hunk 2]
```diff
@@ -563,20 +606,21 @@ func (dc *DisruptionController) enqueueStalePodDisruptionCleanup(pod *v1.Pod, d
...
+	pdbs, err := dc.pdbLister.GetPodLabelsDisruptionBudgets(podNamespace, podLabels)
 	if err != nil {
-		klog.V(4).Infof("No PodDisruptionBudgets found for pod %v, PodDisruptionBudget controller will avoid syncing.", pod.Name)
+		klog.V(4).Infof("No PodDisruptionBudgets found for pod labels %s, PodDisruptionBudget controller will avoid syncing.", podLabels.String())
 		return nil
 	}
 
 	if len(pdbs) > 1 {
-		msg := fmt.Sprintf("Pod %q/%q matches multiple PodDisruptionBudgets.  Chose %q arbitrarily.", pod.Namespace, pod.Name, pdbs[0].Name)
+		msg := fmt.Sprintf("Pod labels %q/%q matche multiple PodDisruptionBudgets. Chose %q arbitrarily.", podNamespace, podLabels.String(), pdbs[0].Name)
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go:620]: this adds a typo, and stringification of labels is expensive

File: pkg/controller/disruption/disruption.go [hunk 3]
```diff
@@ -476,33 +486,72 @@ func (dc *DisruptionController) removeDb(obj interface{}) {
...
-	} else {
-		klog.V(4).Infof("addPod %q -> PDB %q", pod.Name, pdb.Name)
-		dc.enqueuePdb(pdb)
-	}
+type podLabelItem struct {
+	encodedLabels string
+	namespace     string
+}
+
+func (dc *DisruptionController) enqueuePod(pod *v1.Pod) {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go:494]: logically, it looks like the only thing we care about here are the labels.

I wonder if three queue levels would make more sense:
1. process pods
  a. add their labels to a map[namespace]{podName, podNamespace, labels}
  b. queue their namespace for processing
2. process namespaces
  a. grab all the sets of labels observed for that namespace
  b. resolve to covering pdbs (logging using the podName/podNamespace if there are multiple pdbs found or no pdbs found)
  c. queue the pdbs for processing
3. process pdbs

as an aside, processing pod-by-pod assumes we will see every pod event... on startup, it seems like we have to do an initial update of all pdbs to cover any pod events we missed... are we doing that?

File: pkg/controller/disruption/disruption.go [hunk 4]
```diff
@@ -476,33 +486,72 @@ func (dc *DisruptionController) removeDb(obj interface{}) {
...
+}
+
+func (dc *DisruptionController) updatePod(oldObj, newObj interface{}) {
+	newPod := newObj.(*v1.Pod)
+	klog.V(4).Infof("updatePod called on pod %q", newPod.Name)
+	dc.enqueuePod(newPod)
+
+	oldPod := oldObj.(*v1.Pod)
+	if labels.Equals(oldPod.Labels, newPod.Labels) {
+		dc.enqueuePod(oldPod)
```
[the remark this answers is not in the document]
Comment (veshij) [reply pkg/controller/disruption/disruption.go:554]: > why are we enqueuing the oldPod in addition to newPod when the labels are equal?

That's a bug, fixed.

File: pkg/controller/disruption/disruption.go (hunk 3 above)
Comment (veshij) [reply pkg/controller/disruption/disruption.go:494]: >I wonder if three queue levels would make more sense
I'm not sure I understand what's the benefit of the suggested approach.

>it seems like we have to do an initial update of all pdbs to cover any pod events we missed... are we doing that?
More or less.
On startup we enqueue all pdbs to DisruptionController.queue, on top of that we also enqueue all unique pod labels in labelUpdateQueue and eventually converge all pdbs which match those labels.

**Pull request closed without merging**